contracts/components/staking/SlashingController.sol

```diff
@@ -0,0 +1,438 @@
+        SlashPenalty[] calldata __slashPenalties
+    ) public initializer {
+        __AccessManaged_init(__manager);
+        __Routed_init(__router);
+        __UUPSUpgradeable_init();
+
+        _setSlashingExecutor(__executor);
+        _setStakingParametersManager(__stakingParameters);
+        _setDepositAmount(__depositAmount);
+        _setSlashPercentToProposer(__slashPercentToProposer);
+        _setSlashPenalties(__slashPenaltyIds, __slashPenalties);
+
+        if (__depositToken == address(0)) revert ZeroAddress("__depositToken");
+        depositToken = IERC20(__depositToken);
+
+        // UNDEFINED --> CREATED
+        uint256[] memory afterUndefined = new uint256[](1);
+        afterUndefined[0] = uint256(SlashStates.CREATED);
+        _configureState(uint256(SlashStates.UNDEFINED), afterUndefined);
+
+        // CREATED --> DISMISSED, REJECTED or IN_REVIEW
+        uint256[] memory afterCreated = new uint256[](3);
+        afterCreated[0] = uint256(SlashStates.DISMISSED);
+        afterCreated[1] = uint256(SlashStates.REJECTED);
+        afterCreated[2] = uint256(SlashStates.IN_REVIEW);
+        _configureState(uint256(SlashStates.CREATED), afterCreated);
+
+        // IN_REVIEW --> REVIEWED or REVERTED
+        uint256[] memory afterInReview = new uint256[](2);
+        afterInReview[0] = uint256(SlashStates.REVIEWED);
+        afterInReview[1] = uint256(SlashStates.REVERTED);
+        _configureState(uint256(SlashStates.IN_REVIEW), afterInReview);
+
+        // REVIEWED --> EXECUTED or REVERTED
+        uint256[] memory afterReviewed = new uint256[](2);
+        afterReviewed[0] = uint256(SlashStates.EXECUTED);
+        afterReviewed[1] = uint256(SlashStates.REVERTED);
+        _configureState(uint256(SlashStates.REVIEWED), afterReviewed);
+    }
+
+    // Proposal LifeCycle
+    /**
+     * @notice Creates a slash proposal pointing to a slashable subject. To do so, the proposer must provide a FORT deposit and present evidence.
+     * @param _subjectType type of the subject.
+     * @param _subjectId ERC721 registry id of the stake subject.
+     * @param _penaltyId if of the SlashPenalty to inflict upon the subject if the proposal goes through.
+     * @param _evidence IPFS hashes of the evidence files, proof of the subject being slash worthy.
+     * @return proposalId the proposal identifier.
+     */
+    function proposeSlash(
+        uint8 _subjectType,
+        uint256 _subjectId,
+        bytes32 _penaltyId,
+        string[] calldata _evidence
+    ) external onlyValidSlashPenaltyId(_penaltyId) onlyValidSubjectType(_subjectType) returns (uint256 proposalId) {
+        if (!stakingParameters.isRegistered(_subjectType, _subjectId)) revert NonRegisteredSubject(_subjectType, _subjectId);
+        if (stakingParameters.totalStakeFor(_subjectType, _subjectId) == 0) revert ZeroAmount("subject stake");
+        Proposal memory slashProposal = Proposal(_subjectId, msg.sender, _penaltyId, _subjectType);
```

Comment:
SECURITY:
this contract supports the forwarder. it should use `_msgSender()` instead of `msg.sender`. 
(multiple occurrences)

Reply:
Fixed in b6ed756